### ISSUE TITLE
In Azure (ARO) the egress-IPs are added to the backend pool of the ingress load balancer which causes intermittant failures

### ISSUE THREAD

**Issue by paulczar** [redhat-cop/egressip-ipam-operator#86]
You can see the two Egress IPs created for namespaces as the bottom two addresses.

![Screenshot 2021-12-17 132713](https://user-images.githubusercontent.com/2488346/146597567-0cd32ce3-1c9c-4211-8335-188e360d6418.png)

When I curl the openshift console I get intermittent failures, if I remove the namespaces (and thus the egress ips) they disappear from the LB backend pool and all connections work.

example of failures:
```
while true; do curl -s https://console-openshift-console.apps.xxxx.eastus.aroapp.io > /dev/null && echo OK || echo bad; done
bad
OK
OK
bad
OK
OK
bad
OK
bad
bad
bad
bad
```

**Comment by raffaelespazzoli** [comment]
egress IPs are indeed added to the backend pool (https://github.com/redhat-cop/egressip-ipam-operator/blob/448cd8982f0ed98dc3d628d0d806ec0b320fc722/controllers/egressipam/azure/azure.go#L471). Apparently without that setting the outbound traffic is not routed correctly. I frankly never understood why. I considered it to be an Azure quirk.
Is there a better way to do this configuration?

**Comment by raffaelespazzoli** [comment]
@amelie1979  any comment on the above?

**Comment by paulczar** [comment]
Ahh okay, I think I know what's going on then!   

By default ARO has an outbound-rule-v4 attached to the LB that allows traffic out.   However this causes issues with egressip-ipam-operator because a LB can't have outbound rules for backend hosts with more than one IP (see #74). 

To use the operator, the ARO admin has to delete this outbound rule and provide an alternative route out (usually via a firewall see the first section of [here](https://mobb.ninja/docs/aro/egress-ipam-operator).   

With the ^ steps above done, the egress-ips should route out via the firewall appropriately and not need to be part of the pool, given the operator doesn't work when the outbound-rule is in place, I think its safe to remove line 741.

**Comment by paulczar** [comment]
@amelie1979 @raffaelespazzoli Any thoughts on the best way to progress this issue ? I believe this is a blocker for any ARO users.

**Comment by raffaelespazzoli** [comment]
so, usually this is a two phase process:
1. understanding what the configuration should look like in OCP/Azure.
2. design how to automate that configuration via the operator. 
I don't understand #1. Let me try to take a stab at it. The way the operator works today is fine for IPI, but it may not work for another types of Azure deployment. The egress load balancer should can be different.
Is this different egress load balancer value discoverable or is it an input? is it the same for every node..... I am not familiar with Azure, I don't know the answers to all of these questions.
In other words, if you could for this research and come up with a design, then we can look at implementing it in the operator.

**Comment by paulczar** [comment]
@raffaelespazzoli I believe your correct, unfortunately I don't know enough about the underlying architecture choice differences betwen OCP IPI and ARO, but ARO creates two LBs, one for ingress, one for egress to which each new interface is added.   With a customer providing a custom route the egress is not needed, and in both cases the ingress is.  I am thinking our best option here might actually be to add an option to disable https://github.com/redhat-cop/egressip-ipam-operator/blob/448cd8982f0ed98dc3d628d0d806ec0b320fc722/controllers/egressipam/azure/azure.go#L471.   either as a CLI flag for the operator or as a extra CRD option.

**Comment by paulczar** [comment]
I rebuild the Operator without line 471 and it works for an ARO cluster with egress lockdown without causing the intermittant ingress failures.

**Comment by raffaelespazzoli** [comment]
ok, so perhaps we could add an annotation specifying whether the IPConfiguration should be initialized with the egress load balancer... But Is till think we should understand what the correct design should be for Azure.

**Comment by paulczar** [comment]
I can confirm that our custom has updated their deployment to use my modification and are happy with the results, so its now officially tested in production :)

I agree it would be great to get someone to provide specific details on the correct design for OSD on Azure and if its different to ARO, the same for ARO.

**Comment by raffaelespazzoli** [comment]
wonderful, I'm working on getting some expert eyes on this.

**Comment by paulczar** [comment]
Looking at this previous issue #61 it appears that the install time configuration `publish: internal` might determine the OCP behavior here that exposes this bug/issue.

```
That is correct. In the cluster where we got it working we did not use publish internal. I guess that could be the issue. Is there a way @amelie1979 could test the operator using the publish internal option?
```

I wonder if we can look this up in the openshift config and use that to determine if we add the LB pools or not, if we can't get visibility to that config option, then an annotation would seem to be the way to go.

**Comment by raffaelespazzoli** [comment]
@paulczar let's assume we can read the configuration, what would the rule be?
`publish: internal` -> don't add the egress load balancer to the IPConfiguration?

**Comment by paulczar** [comment]
Further digging on both a private and public ARO cluster show the same behavior here (plus the same failures because of the backend pool for an outbound rule).  I don't have a non-ARO cluster to validate on.

I think your suggested rule would solve the problem for ARO, and likely for some OCP clusters that are created to be internal.
